File: src/main/java/net/softbell/bsh/domain/ControlModeRule.java
package net.softbell.bsh.domain;

import lombok.AllArgsConstructor;
import lombok.Getter;

/**
 * @Author : Bell(bell@softbell.net)
 * @Description : 제어 모드 비트플래그 자료형
 */
@AllArgsConstructor
@Getter
public enum ControlModeRule
{
	INFO_SYNC_DISABLE("INFO_SYNC_DISABLE", 				(byte)0b00000001),
	AUTO_CONTROL_DISABLE("AUTO_CONTROL_DISABLE", 		(byte)0b00000010),
	MANUAL_CONTROL_DISABLE("MANUAL_CONTROL_DISABLE", 	(byte)0b00000100);

	private String value;
    private Byte code;
    
    public static ControlModeRule ofLegacyCode(Byte legacyCode)
    {
    	for (ControlModeRule authStatusRule : values())
            if (authStatusRule.getCode() == legacyCode)
                return authStatusRule;
    	
    	return null;
    }
}

File: src/main/kotlin/net/softbell/bsh/domain/ControlModeRule.kt
package net.softbell.bsh.domain

/**
 * @author : Bell(bell@softbell.net)
 * @description : 제어 모드 비트플래그 자료형
 */
enum class ControlModeRule {
    UNKNOWN { // 미확인
        override val value: String
            get() = "UNKNOWN"
        override val code: Byte
            get() = -1
    },
    INFO_SYNC_DISABLE {
        override val value: String
            get() = "INFO_SYNC_DISABLE"
        override val code: Byte
            get() = 1
    },
    AUTO_CONTROL_DISABLE {
        override val value: String
            get() = "AUTO_CONTROL_DISABLE"
        override val code: Byte
            get() = 2
    },
    MANUAL_CONTROL_DISABLE {
        override val value: String
            get() = "MANUAL_CONTROL_DISABLE"
        override val code: Byte
            get() = 4
    };

    abstract val value: String
    abstract val code: Byte

    companion object {
        fun ofLegacyCode(legacyCode: Byte): ControlModeRule {
            for (rule in values())
                if (rule.code == legacyCode)
                    return rule
            return UNKNOWN
        }
    }
}